File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque

def make_graph(words):
    graph = dict()
    for w in words:
        graph[w] = []
    for i in range(len(words)):
        word = words[i]
        for j in range(i+1, len(words)):
            next_word = words[j]
            n = 0
            for k in range(len(words[0])):
                if word[k] == next_word[k]:
                    n += 1
            if n == len(words[0])-1:
                graph[word].append(next_word)
                graph[next_word].append(word)
    return graph

def bfs(graph, start, target):
    queue = deque([(start, 0)])
    visited = [start]
    while queue:
        p, cnt = queue.popleft()
        for w in graph[p]:
            if w == target:
                return cnt+1
            if w not in visited:
                queue.append((w, cnt+1))
                visited.append(w)

def solution(begin, target, words):
    if target not in words:
        return 0

    # 한 글자만 다른 단어끼리 연결해서 그래프 만들기
    words.append(begin)
    graph = make_graph(words)
    answer = bfs(graph, begin, target)

    return answer

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;


class Solution {
    
    static class WordCnt {
        String word;
        int cnt;
        
        public WordCnt(String word, int cnt) {
            this.word = word;
            this.cnt = cnt;
        }
    }
    
    public int solution(String begin, String target, String[] words) {
        Set<String> visited = new HashSet<>();
        
        Queue<WordCnt> queue = new LinkedList<>();
        queue.offer(new WordCnt(begin, 0));
        visited.add(begin);
        
        while (!queue.isEmpty()) {
            WordCnt wordCnt = queue.poll();
            
            String word = wordCnt.word;
            int cnt = wordCnt.cnt;
            
            if (word.equals(target)) {
                return cnt;
            }
            
            for (int i = 0; i < words.length; i++) {
                String w = words[i];
                
                if (check(word, w) && !visited.contains(w)) {
                    queue.offer(new WordCnt(w, cnt+1));
                    visited.add(w);
                }
            }
        }
        
        return 0;
    }
    
    public boolean check(String s1, String s2) {
        int n = s1.length();
        int cnt = 0;
        for (int i = 0; i < n; i++) {
            if (s1.charAt(i) != s2.charAt(i)) {
                cnt++;
                if (cnt >= 2) {
                    return false;
                }
            }
        }
        return true;
    }
}